### PR TITLE
Force DNI fields required if associated Country needs it

### LINKED ISSUE #9891: [BOOM-3848] DNI mandatory for Italy, but not for other countries
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-3848).

- _**Reporter:**_ Rainbow
- _**Created at:**_ Fri, 15 Sep 2017 11:25:58 +0200

<p>I have 2 countries: France & Italy</p>

<p>For Italy, the DNI is required, I checked it in Customers > Addresses > Set required fields for this section<br/>
 I add also the DNI field in International > Locations > Countries > Italy</p>

<p>If I try to make an order with an address in France, I can't because the DNI field isn't completed</p>

<p>The DNI field should be mandatory only for countries in which it has been added (address format) AND if DNI is checked as required in Customers > Addresses</p>
        
* How to reproduce the issue ?

N/A

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Some country needs the dni field in their customer addresses, but you can't force it on every countries So the address validation needs to check this field according to the country settings and for the required field if needed.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9891
| How to test?  | Go to International > Locations > Countries, select a country and set "Do you need a tax identification number?" to true, and you need to add "dni" in the address format or you won't be able to see (nor validate) the field. Try to enter an address in this country without dni you are blocked. When you try in another country everything is fine. 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11726)

### CODE REVIEW

File: classes/Address.php [hunk 1]
```diff
@@ -327,6 +327,34 @@ public static function isCountryActiveById($id_address)
...
+     * @inheritdoc
+     */
+    public function validateField($field, $value, $id_lang = null, $skip = array(), $human_errors = false)
+    {
+        if (!($error = parent::validateField($field, $value, $id_lang, $skip, $human_errors))) {
+            return $error;
+        }
+
+        //Special validation for dni, check if the country needs it
+        if ('dni' === $field) {
```
Comment (PierreRambaud) [inline classes/Address.php]:
```php
if ( 'dni' !== $field) {
  return true;
}
```

File: classes/Address.php [hunk 2]
```diff
@@ -327,6 +327,34 @@ public static function isCountryActiveById($id_address)
...
+            $result = (bool) Db::getInstance(_PS_USE_SQL_SLAVE_)->getValue('
+			SELECT c.`need_identification_number`
+			FROM `' . _DB_PREFIX_ . 'country` c
+			WHERE c.`id_country` = ' . (int) $this->id_country);
+
+            if ($result && Tools::isEmpty($value)) {
+                if ($human_errors) {
+                    return $this->trans('The %s field is required.', array($this->displayFieldName($field, get_class($this))), 'Admin.Notifications.Error');
+                } else {
```
Comment (PierreRambaud) [inline classes/Address.php]: Useless else.

File: classes/Address.php [hunk 3]
```diff
@@ -327,6 +327,34 @@ public static function isCountryActiveById($id_address)
         return Cache::retrieve($cache_id);
     }
 
+    /**
+     * @inheritdoc
+     */
+    public function validateField($field, $value, $id_lang = null, $skip = array(), $human_errors = false)
+    {
+        if (!($error = parent::validateField($field, $value, $id_lang, $skip, $human_errors))) {
```
Comment (PierreRambaud) [inline classes/Address.php]: If no errors, return error?

File: classes/Address.php (hunk 1 above)
Comment (jolelievre) [reply classes/Address.php]: done

File: classes/Address.php (hunk 3 above)
Comment (jolelievre) [reply classes/Address.php]: nice catch!!

File: classes/Address.php (hunk 2 above)
Comment (jolelievre) [reply classes/Address.php]: done

**Review by PierreRambaud — DISMISSED** [review]

**Comment by marionf** [comment]
@jolelievre 

I have set "Do you need a tax identification number?" to Yes for Italy and added "dni" in the address format
When I try to add an address in Italy from the FO, I have this Exception

capture d ecran_836

capture d ecran_835

Also, I can't complete the order while the identification number field is well filled

capture d ecran_837

If I edit this address the identification number is well filled

capture d ecran_840

But when I save I have another exception

capture d ecran_839

If I fill the company field (optional), I will have the same exception for the next optional fields: vat number, address2, other etc ...

**Review by jolelievre — COMMENTED** [review]
@PierreRambaud seems good to me, but as an author I can't approve it
either you can do it or someone from @PrestaShop/prestashop-core-developers ?

**Review by matks — APPROVED** [review]

**Comment by eternoendless** [comment]
Thank you @jolelievre and @PierreRambaud

**Pull request merged**